{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# init\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from tqdm import tqdm\n",
    "# import nltk\n",
    "# nltk.download()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import glob\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "from nltk.tokenize import word_tokenize\n",
    "from nltk.corpus import stopwords"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import math\n",
    "from sklearn.feature_extraction.text import CountVectorizer\n",
    "from sklearn.feature_extraction.text import TfidfVectorizer\n",
    "from sklearn.preprocessing import LabelBinarizer\n",
    "from sklearn.preprocessing import LabelEncoder\n",
    "from scipy.sparse import spdiags"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.cluster import KMeans\n",
    "# from spherecluster import SphericalKMeans\n",
    "from matplotlib import pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.metrics import silhouette_score\n",
    "from sklearn.metrics.cluster import contingency_matrix\n",
    "from sklearn.metrics import adjusted_mutual_info_score"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Import Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import pubmed\n",
    "pubmed = []\n",
    "docList = glob.glob(os.path.join(os.getcwd(), \"Datasets/pubmed/\", \"*.txt\"))\n",
    "\n",
    "for docPath in tqdm(docList):\n",
    "    # get doc file name\n",
    "    docName = os.path.basename(docPath).split('.')[0]\n",
    "    className = docName[:3]\n",
    "    \n",
    "    with open(docPath, encoding=\"utf8\") as doc:\n",
    "        # insert [class, docs, feature]\n",
    "        pubmed.append([className, docName, doc.read().replace('\\n', ' ')])\n",
    "\n",
    "# make dataframe\n",
    "dataframe = pd.DataFrame(data=pubmed, columns=['class', 'document', 'feature']) \n",
    "\n",
    "# export pubmed raw\n",
    "dataframe.to_csv('pubmed/raw.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import scopus\n",
    "scopus = []\n",
    "docList = glob.glob(os.path.join(os.getcwd(), \"Datasets/scopus/\", \"*\"))\n",
    "\n",
    "for docPath in tqdm(docList):\n",
    "    # get doc file name\n",
    "    docName = os.path.basename(docPath)\n",
    "    className = docName.split('-')[0]\n",
    "    \n",
    "    with open(docPath, encoding=\"utf8\") as doc:\n",
    "        # insert [class, docs, feature]\n",
    "        scopus.append([className, docName, doc.read().replace('\\n', ' ')])\n",
    "\n",
    "# make dataframe\n",
    "dataframe = pd.DataFrame(data=scopus, columns=['class', 'document', 'feature']) \n",
    "\n",
    "# export pubmed raw\n",
    "dataframe.to_csv('scopus/raw.csv', index=False)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Dataframe Raw"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "613"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read csv\n",
    "pubmedRaw = pd.read_csv('pubmed/raw.csv')\n",
    "scopusRaw = pd.read_csv('scopus/raw.csv')\n",
    "\n",
    "# get feature\n",
    "pubmedFeatures = pubmedRaw.loc[:, 'feature']\n",
    "scopusFeatures = scopusRaw.loc[:, 'feature']\n",
    "scopusFeatures"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Preprocesing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# cleaning\n",
    "def cleaning(features):\n",
    "    result = []\n",
    "    for feature in features:\n",
    "        regex = re.sub(r'[^a-zA-Z\\s]', '', feature)\n",
    "        result.append(regex)\n",
    "    return result\n",
    "\n",
    "# case folding\n",
    "def caseFolding(features):\n",
    "    result = []\n",
    "    for feature in features:\n",
    "        lower = feature.lower()\n",
    "        result.append(lower)\n",
    "    return result\n",
    "\n",
    "# tokenization\n",
    "def tokenization(features):\n",
    "    result = []\n",
    "    for feature in features:\n",
    "        token = word_tokenize(feature)\n",
    "        result.append(token)\n",
    "    return result\n",
    "\n",
    "# stopwords removal\n",
    "def stopWords(features):\n",
    "    result = []\n",
    "    stopWords = set(stopwords.words('english'))\n",
    "    for token in features:\n",
    "        cleanedFeature = [feature for feature in token if feature not in stopWords]\n",
    "        result.append(cleanedFeature)\n",
    "    return result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "# preprocessing\n",
    "def preprocessing(features):\n",
    "    clean = cleaning(features)\n",
    "    case = caseFolding(clean)\n",
    "    token = tokenization(case)\n",
    "    preprocessed = stopWords(token)\n",
    "    return preprocessed\n",
    "    \n",
    "pubmedPreprocessed = preprocessing(pubmedFeatures)\n",
    "scopusPreprocessed = preprocessing(scopusFeatures)\n",
    "# print(scopusPreprocessed[0])\n",
    "\n",
    "# export pubmed clean\n",
    "for i in range(len(pubmedPreprocessed)):\n",
    "    pubmedRaw.loc[i, 'feature'] = ' '.join(pubmedPreprocessed[i])\n",
    "pubmedRaw.to_csv('pubmed/clean.csv', index=False)\n",
    "\n",
    "# export scopus clean\n",
    "for i in range(len(scopusPreprocessed)):\n",
    "    scopusRaw.loc[i, 'feature'] = ' '.join(scopusPreprocessed[i])\n",
    "scopusRaw.to_csv('scopus/clean.csv', index=False)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Daraframe Clean"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "473"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read csv\n",
    "pubmedClean = pd.read_csv('pubmed/clean.csv')\n",
    "scopusClean = pd.read_csv('scopus/clean.csv')\n",
    "\n",
    "# get feature\n",
    "pubmedFeatures = pubmedClean.loc[:, 'feature']\n",
    "scopusFeatures = scopusClean.loc[:, 'feature']\n",
    "scopusClean"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Feature Forming"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from textblob import TextBlob\n",
    "\n",
    "def featureForming(features, dataframe):\n",
    "    # BOAW\n",
    "    dataframe.rename(columns={'feature': 'BOAW'}, inplace=True)\n",
    "    for i in tqdm(range(len(features))):\n",
    "        # BON\n",
    "        dataframe.loc[i, 'BON'] = ' '.join(TextBlob(features[i]).noun_phrases)\n",
    "        # BONA\n",
    "        dataframe.loc[i, 'BONA'] = ' '.join([word for (word, tag) in TextBlob(features[i]).tags if tag[:2]=='NN' or tag[:2]=='JJ'])\n",
    "\n",
    "featureForming(pubmedFeatures, pubmedClean)\n",
    "featureForming(scopusFeatures, scopusClean)\n",
    "# print(scopusClean)\n",
    "\n",
    "pubmedClean.to_csv('pubmed/formed.csv', index=False)\n",
    "scopusClean.to_csv('scopus/formed.csv', index=False)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Dataframe Formed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0             CONCRETE\n",
       "1             CONCRETE\n",
       "2             CONCRETE\n",
       "3             CONCRETE\n",
       "4             CONCRETE\n",
       "             ...      \n",
       "2795    TECTONICPLATES\n",
       "2796    TECTONICPLATES\n",
       "2797    TECTONICPLATES\n",
       "2798    TECTONICPLATES\n",
       "2799    TECTONICPLATES\n",
       "Name: class, Length: 2800, dtype: object"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read csv\n",
    "pubmedFormed = pd.read_csv('pubmed/formed.csv')\n",
    "scopusFormed = pd.read_csv('scopus/formed.csv')\n",
    "\n",
    "# get feature\n",
    "# pubmed\n",
    "pubmedClass = pubmedFormed.loc[:, 'class']\n",
    "pubmed_boaw = pubmedFormed.loc[:, 'BOAW']\n",
    "pubmed_bon = pubmedFormed.loc[:, 'BON']\n",
    "pubmed_bona = pubmedFormed.loc[:, 'BONA']\n",
    "\n",
    "# scopus\n",
    "scopusClass = scopusFormed.loc[:, 'class']\n",
    "scopus_boaw = scopusFormed.loc[:, 'BOAW']\n",
    "scopus_bon = scopusFormed.loc[:, 'BON']\n",
    "scopus_bona = scopusFormed.loc[:, 'BONA']\n",
    "scopusClass"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Feature Selection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# SFS\n",
    "def symbolic(features, classes):\n",
    "    # get tf weights\n",
    "    tfVec = CountVectorizer()\n",
    "    tf = tfVec.fit_transform(features)\n",
    "\n",
    "    # define weights into dataframe\n",
    "    featureName = tfVec.get_feature_names_out()\n",
    "    featureWeight = tf.todense()\n",
    "    df = pd.DataFrame(featureWeight, columns=featureName)\n",
    "    df.loc[:, 'class'] = classes\n",
    "    # return df\n",
    "\n",
    "    # count mean & standard deviation\n",
    "    mean = df.groupby('class').mean().reset_index()\n",
    "    std = df.groupby('class').std().reset_index()\n",
    "\n",
    "    totalSm = []\n",
    "    for col in tqdm(mean.columns):\n",
    "        if col != 'class':\n",
    "            # count interval per features\n",
    "            interval = []\n",
    "            for row in range(len(mean)):\n",
    "                meanA = mean.loc[row, col]\n",
    "                stdA = std.loc[row, col]\n",
    "                interval.append([meanA - stdA, meanA + stdA])\n",
    "            \n",
    "            # count similarity per feature\n",
    "            similarity = 0\n",
    "            for itvA in interval:\n",
    "                # count similarity per class\n",
    "                smClass = []\n",
    "                for itvB in interval:\n",
    "                    if itvA != itvB:\n",
    "                        # check if interval B equals 0\n",
    "                        sm = 0\n",
    "                        if (itvB[1] - itvB[0]) != 0:\n",
    "                            sm = (min(itvA[1], itvB[1]) - max(itvA[0], itvB[0])) / (itvB[1] - itvB[0])\n",
    "                        smClass.append(sm)\n",
    "\n",
    "                # count total similarity\n",
    "                similarity += np.sum(smClass)\n",
    "            totalSm.append(similarity)\n",
    "\n",
    "    # count average total similarity\n",
    "    avgTotalSm = np.mean(totalSm)\n",
    "\n",
    "    # select feature that totalSm > avgTotalSm\n",
    "    selected = []\n",
    "    for i in range(len(totalSm)):\n",
    "        if totalSm[i] > avgTotalSm:\n",
    "            selected.append(df.columns[i])\n",
    "    # return len(selected)\n",
    "    return selected\n",
    "\n",
    "# sfx = symbolic(scopus_boaw, scopusClass)\n",
    "# print(len(sfx))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# TF-RF\n",
    "def relevance(features, classes):\n",
    "    # get tf weights\n",
    "    tfVec = CountVectorizer()\n",
    "    tf = tfVec.fit_transform(features)\n",
    "\n",
    "    # get classes weights\n",
    "    clsVec = LabelBinarizer()\n",
    "    cls = clsVec.fit_transform(classes)\n",
    "    # return cls, cls.shape\n",
    "\n",
    "    # if has only 1 class, add negative class\n",
    "    if cls.shape[1] == 1:\n",
    "        cls = np.append(1 - cls, cls, axis=1)\n",
    "\n",
    "    # count class contain feature (1 if contain, 0 if not, then convert to float)\n",
    "    # featureDoc = ((cls.T * tf) > 0).astype(np.float64)\n",
    "    featureDoc = (cls.T * tf).astype(np.float64)\n",
    "    # return featureDoc.shape\n",
    "\n",
    "    for classC in range(cls.shape[1]):\n",
    "        # featureDoc.sum(axis=0) - featureDoc[classC] is the sum of all rows except classC\n",
    "        featureDoc[classC] /= np.maximum(1., featureDoc.sum(axis=0) - featureDoc[classC])\n",
    "\n",
    "    # count rf\n",
    "    rf = np.mean(np.log2(2 + featureDoc, out=featureDoc), axis=0)\n",
    "\n",
    "    # count tf-rf & transform to sparse matrix\n",
    "    totalFeature = rf.shape[0]\n",
    "    tfRf = tf * spdiags(rf, 0, totalFeature, totalFeature)\n",
    "\n",
    "    # define weights into dataframe\n",
    "    featureName = tfVec.get_feature_names_out()\n",
    "    featureWeight = tfRf.todense()\n",
    "    df = pd.DataFrame(featureWeight, columns=featureName)\n",
    "    df.loc[:, 'class'] = classes\n",
    "    # return df\n",
    "\n",
    "    # count weight by class\n",
    "    mean = df.groupby('class').mean()\n",
    "    result = (mean/mean.max()).sum()\n",
    "    # return result\n",
    "\n",
    "    # select minimum weighted features\n",
    "    selected = []\n",
    "    for i in tqdm(range(len(result))):\n",
    "        if result[i] > np.mean(result):\n",
    "            selected.append(df.columns[i])\n",
    "    # return len(selected)\n",
    "    return selected\n",
    "\n",
    "# rfx = relevance(pubmed_bon, pubmedClass)\n",
    "# print(len(rfx))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Implementation\n",
    "# Symbolic Relevance\n",
    "def symbolic_relevance(features, classes):\n",
    "    # get sfs\n",
    "    sfs = symbolic(features, classes)\n",
    "    # get tf-rf\n",
    "    tfrf = relevance(features, classes)\n",
    "\n",
    "    # intersect features\n",
    "    selected = list(np.intersect1d(sfs, tfrf))\n",
    "    # print(len(selected))\n",
    "\n",
    "    # update features\n",
    "    token = tokenization(features)\n",
    "    result = []\n",
    "    for row in tqdm(token):\n",
    "        words = [word for word in row if word in selected]\n",
    "        result.append(words)\n",
    "    return result\n",
    "\n",
    "# srx = symbolic_relevance(scopus_bona, scopusClass)\n",
    "# print(srx)\n",
    "\n",
    "# SFS\n",
    "def symbolic_feature_selection(features, classes):\n",
    "    # get sfs\n",
    "    selected = symbolic(features, classes)\n",
    "    # print(len(selected))\n",
    "\n",
    "    # update features\n",
    "    token = tokenization(features)\n",
    "    result = []\n",
    "    for row in tqdm(token):\n",
    "        words = [word for word in row if word in selected]\n",
    "        result.append(words)\n",
    "    return result\n",
    "\n",
    "# sfsx = sfs(scopus_bona, scopusClass)\n",
    "# print(sfsx)\n",
    "\n",
    "# TF-RF\n",
    "def term_frequency_relevance_frequency(features, classes):\n",
    "    # get tfrf\n",
    "    selected = relevance(features, classes)\n",
    "    # print(len(selected))\n",
    "\n",
    "    # update features\n",
    "    token = tokenization(features)\n",
    "    result = []\n",
    "    for row in tqdm(token):\n",
    "        words = [word for word in row if word in selected]\n",
    "        result.append(words)\n",
    "    return result\n",
    "\n",
    "# tfrfx = tfrf(scopus_bona, scopusClass)\n",
    "# print(tfrfx)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 21873/21873 [00:04<00:00, 5264.37it/s]\n",
      "100%|██████████| 21872/21872 [00:02<00:00, 9076.93it/s]\n",
      "100%|██████████| 4000/4000 [00:20<00:00, 197.02it/s]\n",
      "100%|██████████| 17596/17596 [00:03<00:00, 4831.41it/s]\n",
      "100%|██████████| 17595/17595 [00:01<00:00, 9965.60it/s] \n",
      "100%|██████████| 4000/4000 [00:13<00:00, 301.94it/s]\n",
      "100%|██████████| 18358/18358 [00:03<00:00, 4991.61it/s]\n",
      "100%|██████████| 18357/18357 [00:01<00:00, 9806.97it/s]\n",
      "100%|██████████| 4000/4000 [00:14<00:00, 277.44it/s]\n",
      "100%|██████████| 32980/32980 [00:09<00:00, 3464.88it/s]\n",
      "100%|██████████| 32979/32979 [00:04<00:00, 7592.20it/s]\n",
      "100%|██████████| 2800/2800 [00:22<00:00, 122.26it/s]\n",
      "100%|██████████| 26341/26341 [00:07<00:00, 3533.22it/s]\n",
      "100%|██████████| 26340/26340 [00:03<00:00, 8118.46it/s]\n",
      "100%|██████████| 2800/2800 [00:11<00:00, 240.72it/s]\n",
      "100%|██████████| 27044/27044 [00:08<00:00, 3364.42it/s]\n",
      "100%|██████████| 27043/27043 [00:03<00:00, 8019.75it/s]\n",
      "100%|██████████| 2800/2800 [00:13<00:00, 210.65it/s]\n",
      "100%|██████████| 21873/21873 [00:04<00:00, 4749.09it/s]\n",
      "100%|██████████| 4000/4000 [00:26<00:00, 153.78it/s]\n",
      "100%|██████████| 17596/17596 [00:03<00:00, 4817.26it/s]\n",
      "100%|██████████| 4000/4000 [00:15<00:00, 261.23it/s]\n",
      "100%|██████████| 18358/18358 [00:03<00:00, 4789.68it/s]\n",
      "100%|██████████| 4000/4000 [00:17<00:00, 234.79it/s]\n",
      "100%|██████████| 32980/32980 [00:09<00:00, 3477.71it/s]\n",
      "100%|██████████| 2800/2800 [00:30<00:00, 90.66it/s] \n",
      "100%|██████████| 26341/26341 [00:07<00:00, 3532.17it/s]\n",
      "100%|██████████| 2800/2800 [00:15<00:00, 180.79it/s]\n",
      "100%|██████████| 27044/27044 [00:07<00:00, 3514.24it/s]\n",
      "100%|██████████| 2800/2800 [00:16<00:00, 165.33it/s]\n",
      "100%|██████████| 21872/21872 [00:02<00:00, 8097.26it/s]\n",
      "100%|██████████| 4000/4000 [00:41<00:00, 97.04it/s] \n",
      "100%|██████████| 17595/17595 [00:01<00:00, 9705.17it/s] \n",
      "100%|██████████| 4000/4000 [00:23<00:00, 172.47it/s]\n",
      "100%|██████████| 18357/18357 [00:01<00:00, 9206.72it/s]\n",
      "100%|██████████| 4000/4000 [00:25<00:00, 157.09it/s]\n",
      "100%|██████████| 32979/32979 [00:04<00:00, 7355.56it/s]\n",
      "100%|██████████| 2800/2800 [00:48<00:00, 57.15it/s]\n",
      "100%|██████████| 26340/26340 [00:03<00:00, 8010.92it/s]\n",
      "100%|██████████| 2800/2800 [00:25<00:00, 108.98it/s]\n",
      "100%|██████████| 27043/27043 [00:03<00:00, 8337.56it/s]\n",
      "100%|██████████| 2800/2800 [00:28<00:00, 98.17it/s] \n"
     ]
    }
   ],
   "source": [
    "# Symbolic Relevance\n",
    "# pubmed\n",
    "pubmed_boaw_sr = symbolic_relevance(pubmed_boaw, pubmedClass)\n",
    "pubmed_bon_sr = symbolic_relevance(pubmed_bon, pubmedClass)\n",
    "pubmed_bona_sr = symbolic_relevance(pubmed_bona, pubmedClass)\n",
    "\n",
    "# scopus\n",
    "scopus_boaw_sr = symbolic_relevance(scopus_boaw, scopusClass)\n",
    "scopus_bon_sr = symbolic_relevance(scopus_bon, scopusClass)\n",
    "scopus_bona_sr = symbolic_relevance(scopus_bona, scopusClass)\n",
    "\n",
    "# SFS\n",
    "# pubmed\n",
    "pubmed_boaw_sfs = symbolic_feature_selection(pubmed_boaw, pubmedClass)\n",
    "pubmed_bon_sfs = symbolic_feature_selection(pubmed_bon, pubmedClass)\n",
    "pubmed_bona_sfs = symbolic_feature_selection(pubmed_bona, pubmedClass)\n",
    "\n",
    "# scopus\n",
    "scopus_boaw_sfs = symbolic_feature_selection(scopus_boaw, scopusClass)\n",
    "scopus_bon_sfs = symbolic_feature_selection(scopus_bon, scopusClass)\n",
    "scopus_bona_sfs = symbolic_feature_selection(scopus_bona, scopusClass)\n",
    "\n",
    "# TF-RF\n",
    "# pubmed\n",
    "pubmed_boaw_tfrf = term_frequency_relevance_frequency(pubmed_boaw, pubmedClass)\n",
    "pubmed_bon_tfrf = term_frequency_relevance_frequency(pubmed_bon, pubmedClass)\n",
    "pubmed_bona_tfrf = term_frequency_relevance_frequency(pubmed_bona, pubmedClass)\n",
    "\n",
    "# scopus\n",
    "scopus_boaw_tfrf = term_frequency_relevance_frequency(scopus_boaw, scopusClass)\n",
    "scopus_bon_tfrf = term_frequency_relevance_frequency(scopus_bon, scopusClass)\n",
    "scopus_bona_tfrf = term_frequency_relevance_frequency(scopus_bona, scopusClass)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 4000/4000 [00:02<00:00, 1518.06it/s]\n",
      "100%|██████████| 2800/2800 [00:01<00:00, 1727.12it/s]\n"
     ]
    }
   ],
   "source": [
    "# save csv\n",
    "# pubmed\n",
    "pubmedFormed.drop(['BOAW', 'BON', 'BONA'], axis=1, inplace=True)\n",
    "for i in tqdm(range(len(pubmedFormed))):\n",
    "    # BOAW_SR\n",
    "    pubmedFormed.loc[i, 'BOAW_SR'] = ' '.join(pubmed_boaw_sr[i])\n",
    "    # BON_SR\n",
    "    pubmedFormed.loc[i, 'BON_SR'] = ' '.join(pubmed_bon_sr[i])\n",
    "    # BONA_SR\n",
    "    pubmedFormed.loc[i, 'BONA_SR'] = ' '.join(pubmed_bona_sr[i])\n",
    "    \n",
    "    # BOAW_SFS\n",
    "    pubmedFormed.loc[i, 'BOAW_SFS'] = ' '.join(pubmed_boaw_sfs[i])\n",
    "    # BON_SFS\n",
    "    pubmedFormed.loc[i, 'BON_SFS'] = ' '.join(pubmed_bon_sfs[i])\n",
    "    # BONA_SFS\n",
    "    pubmedFormed.loc[i, 'BONA_SFS'] = ' '.join(pubmed_bona_sfs[i])\n",
    "\n",
    "    # BOAW_TFRF\n",
    "    pubmedFormed.loc[i, 'BOAW_TFRF'] = ' '.join(pubmed_boaw_tfrf[i])\n",
    "    # BON_TFRF\n",
    "    pubmedFormed.loc[i, 'BON_TFRF'] = ' '.join(pubmed_bon_tfrf[i])\n",
    "    # BONA_TFRF\n",
    "    pubmedFormed.loc[i, 'BONA_TFRF'] = ' '.join(pubmed_bona_tfrf[i])\n",
    "\n",
    "pubmedFormed.to_csv('pubmed/selected.csv', index=False)\n",
    "\n",
    "# scopus\n",
    "scopusFormed.drop(['BOAW', 'BON', 'BONA'], axis=1, inplace=True)\n",
    "for i in tqdm(range(len(scopusFormed))):\n",
    "    # BOAW_SR\n",
    "    scopusFormed.loc[i, 'BOAW_SR'] = ' '.join(scopus_boaw_sr[i])\n",
    "    # BON_SR\n",
    "    scopusFormed.loc[i, 'BON_SR'] = ' '.join(scopus_bon_sr[i])\n",
    "    # BONA_SR\n",
    "    scopusFormed.loc[i, 'BONA_SR'] = ' '.join(scopus_bona_sr[i])\n",
    "    \n",
    "    # BOAW_SFS\n",
    "    scopusFormed.loc[i, 'BOAW_SFS'] = ' '.join(scopus_boaw_sfs[i])\n",
    "    # BON_SFS\n",
    "    scopusFormed.loc[i, 'BON_SFS'] = ' '.join(scopus_bon_sfs[i])\n",
    "    # BONA_SFS\n",
    "    scopusFormed.loc[i, 'BONA_SFS'] = ' '.join(scopus_bona_sfs[i])\n",
    "\n",
    "    # BOAW_TFRF\n",
    "    scopusFormed.loc[i, 'BOAW_TFRF'] = ' '.join(scopus_boaw_tfrf[i])\n",
    "    # BON_TFRF\n",
    "    scopusFormed.loc[i, 'BON_TFRF'] = ' '.join(scopus_bon_tfrf[i])\n",
    "    # BONA_TFRF\n",
    "    scopusFormed.loc[i, 'BONA_TFRF'] = ' '.join(scopus_bona_tfrf[i])\n",
    "\n",
    "scopusFormed.to_csv('scopus/selected.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>class</th>\n",
       "      <th>document</th>\n",
       "      <th>BOAW_SR</th>\n",
       "      <th>BON_SR</th>\n",
       "      <th>BONA_SR</th>\n",
       "      <th>BOAW_SFS</th>\n",
       "      <th>BON_SFS</th>\n",
       "      <th>BONA_SFS</th>\n",
       "      <th>BOAW_TFRF</th>\n",
       "      <th>BON_TFRF</th>\n",
       "      <th>BONA_TFRF</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ALZ</td>\n",
       "      <td>ALZ1</td>\n",
       "      <td>reduced amounts tissue patients diagnosis comp...</td>\n",
       "      <td>amounts temporal type tissue patients diagnosi...</td>\n",
       "      <td>reduced amounts temporal type tissue patients ...</td>\n",
       "      <td>reduced amounts cortex brain tissue patients d...</td>\n",
       "      <td>amounts temporal cortex type brain tissue pati...</td>\n",
       "      <td>reduced amounts temporal cortex type brain tis...</td>\n",
       "      <td>reduced amounts tissue patients diagnosis comp...</td>\n",
       "      <td>amounts temporal type tissue patients diagnosi...</td>\n",
       "      <td>reduced amounts temporal type tissue patients ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ALZ</td>\n",
       "      <td>ALZ10</td>\n",
       "      <td>criteria diagnosis patients presenting memory ...</td>\n",
       "      <td>criteria type diagnosis patients memory charac...</td>\n",
       "      <td>criteria diagnosis patients memory deficits ch...</td>\n",
       "      <td>criteria diagnosis patients presenting memory ...</td>\n",
       "      <td>criteria type diagnosis patients memory charac...</td>\n",
       "      <td>criteria diagnosis patients memory deficits ch...</td>\n",
       "      <td>criteria diagnosis patients presenting memory ...</td>\n",
       "      <td>criteria type diagnosis patients memory defici...</td>\n",
       "      <td>criteria diagnosis patients memory deficits ch...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ALZ</td>\n",
       "      <td>ALZ100</td>\n",
       "      <td>synthesis cglucose tissue synthesis co cglucos...</td>\n",
       "      <td>synthesis cglucose tissue synthesis co cglucos...</td>\n",
       "      <td>cglucose tissue co cglucose tissue electron ra...</td>\n",
       "      <td>synthesis cglucose tissue synthesis co cglucos...</td>\n",
       "      <td>synthesis cglucose tissue synthesis co cglucos...</td>\n",
       "      <td>cglucose tissue co cglucose tissue electron ra...</td>\n",
       "      <td>synthesis dioxide cglucose tissue synthesis co...</td>\n",
       "      <td>synthesis dioxide production cglucose tissue s...</td>\n",
       "      <td>dioxide production cglucose tissue co producti...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ALZ</td>\n",
       "      <td>ALZ1000</td>\n",
       "      <td>pattern observations thirteen patients dat abi...</td>\n",
       "      <td>pattern type observations patients type dat co...</td>\n",
       "      <td>pattern type thirteen patients type dat abilit...</td>\n",
       "      <td>pattern observations thirteen patients dat abi...</td>\n",
       "      <td>pattern type observations patients type dat co...</td>\n",
       "      <td>pattern type thirteen patients type dat abilit...</td>\n",
       "      <td>pattern observations thirteen patients dat abi...</td>\n",
       "      <td>pattern reading type observations patients typ...</td>\n",
       "      <td>pattern type thirteen patients type dat abilit...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ALZ</td>\n",
       "      <td>ALZ101</td>\n",
       "      <td>blood rate bodies amino acids blood rate bodie...</td>\n",
       "      <td>blood rate bodies amino acids blood rate bodie...</td>\n",
       "      <td>blood rate bodies acids blood rate bodies acid...</td>\n",
       "      <td>cerebral blood rate bodies amino acids cerebra...</td>\n",
       "      <td>cerebral blood rate bodies amino acids cerebra...</td>\n",
       "      <td>cerebral blood rate bodies acids cerebral bloo...</td>\n",
       "      <td>blood rate bodies amino acids blood rate bodie...</td>\n",
       "      <td>blood rate bodies amino acids blood rate bodie...</td>\n",
       "      <td>blood rate bodies acids blood rate bodies acid...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3995</th>\n",
       "      <td>HIV</td>\n",
       "      <td>HIV3996</td>\n",
       "      <td>major genes influence outcome alleles explain ...</td>\n",
       "      <td>major complex genes outcome null alleles assoc...</td>\n",
       "      <td>major genes influence outcome null alleles ass...</td>\n",
       "      <td>major genes influence outcome alleles explain ...</td>\n",
       "      <td>major complex genes outcome null alleles assoc...</td>\n",
       "      <td>major genes influence outcome null alleles ass...</td>\n",
       "      <td>major genes influence outcome alleles explain ...</td>\n",
       "      <td>major complex genes outcome null alleles diver...</td>\n",
       "      <td>major complex genes influence outcome null all...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3996</th>\n",
       "      <td>HIV</td>\n",
       "      <td>HIV3997</td>\n",
       "      <td>cohort patients patients aged clinical examina...</td>\n",
       "      <td>cohort patients course patients clinical exami...</td>\n",
       "      <td>cohort patients course patients years clinical...</td>\n",
       "      <td>cohort patients patients aged clinical examina...</td>\n",
       "      <td>cohort patients course patients clinical exami...</td>\n",
       "      <td>cohort patients course patients years clinical...</td>\n",
       "      <td>cohort patients patients aged years clinical e...</td>\n",
       "      <td>cohort patients course patients clinical exami...</td>\n",
       "      <td>cohort patients course patients years clinical...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3997</th>\n",
       "      <td>HIV</td>\n",
       "      <td>HIV3998</td>\n",
       "      <td>main classifications purpose paper trace descr...</td>\n",
       "      <td>main purpose paper trace describe principal cl...</td>\n",
       "      <td>main classifications purpose paper trace descr...</td>\n",
       "      <td>main classifications purpose paper trace descr...</td>\n",
       "      <td>main purpose paper trace describe principal cl...</td>\n",
       "      <td>main classifications purpose paper trace descr...</td>\n",
       "      <td>main classifications purpose paper trace descr...</td>\n",
       "      <td>main purpose paper trace describe principal cl...</td>\n",
       "      <td>main classifications purpose paper trace descr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3998</th>\n",
       "      <td>HIV</td>\n",
       "      <td>HIV3999</td>\n",
       "      <td>inhibit cell antigen pathway recombinant cell ...</td>\n",
       "      <td>virus inhibit cell antigen pathway recombinant...</td>\n",
       "      <td>virus cell antigen pathway recombinant cell an...</td>\n",
       "      <td>virus inhibit cell antigen pathway recombinant...</td>\n",
       "      <td>virus inhibit cell antigen pathway recombinant...</td>\n",
       "      <td>virus cell antigen pathway recombinant cell an...</td>\n",
       "      <td>inhibit cell antigen pathway interference reco...</td>\n",
       "      <td>immunodeficiency virus inhibit cell antigen pa...</td>\n",
       "      <td>virus cell antigen pathway recombinant cell an...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3999</th>\n",
       "      <td>HIV</td>\n",
       "      <td>HIV4000</td>\n",
       "      <td>predicting progression combined lymphocyte pur...</td>\n",
       "      <td>progression usefulness lymphocyte purpose usef...</td>\n",
       "      <td>progression usefulness lymphocyte investigate ...</td>\n",
       "      <td>predicting progression combined cd lymphocyte ...</td>\n",
       "      <td>progression usefulness cd lymphocyte purpose u...</td>\n",
       "      <td>progression usefulness cd lymphocyte investiga...</td>\n",
       "      <td>predicting progression combined usefulness lym...</td>\n",
       "      <td>progression usefulness lymphocyte purpose usef...</td>\n",
       "      <td>progression usefulness lymphocyte investigate ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4000 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     class document                                            BOAW_SR  \\\n",
       "0      ALZ     ALZ1  reduced amounts tissue patients diagnosis comp...   \n",
       "1      ALZ    ALZ10  criteria diagnosis patients presenting memory ...   \n",
       "2      ALZ   ALZ100  synthesis cglucose tissue synthesis co cglucos...   \n",
       "3      ALZ  ALZ1000  pattern observations thirteen patients dat abi...   \n",
       "4      ALZ   ALZ101  blood rate bodies amino acids blood rate bodie...   \n",
       "...    ...      ...                                                ...   \n",
       "3995   HIV  HIV3996  major genes influence outcome alleles explain ...   \n",
       "3996   HIV  HIV3997  cohort patients patients aged clinical examina...   \n",
       "3997   HIV  HIV3998  main classifications purpose paper trace descr...   \n",
       "3998   HIV  HIV3999  inhibit cell antigen pathway recombinant cell ...   \n",
       "3999   HIV  HIV4000  predicting progression combined lymphocyte pur...   \n",
       "\n",
       "                                                 BON_SR  \\\n",
       "0     amounts temporal type tissue patients diagnosi...   \n",
       "1     criteria type diagnosis patients memory charac...   \n",
       "2     synthesis cglucose tissue synthesis co cglucos...   \n",
       "3     pattern type observations patients type dat co...   \n",
       "4     blood rate bodies amino acids blood rate bodie...   \n",
       "...                                                 ...   \n",
       "3995  major complex genes outcome null alleles assoc...   \n",
       "3996  cohort patients course patients clinical exami...   \n",
       "3997  main purpose paper trace describe principal cl...   \n",
       "3998  virus inhibit cell antigen pathway recombinant...   \n",
       "3999  progression usefulness lymphocyte purpose usef...   \n",
       "\n",
       "                                                BONA_SR  \\\n",
       "0     reduced amounts temporal type tissue patients ...   \n",
       "1     criteria diagnosis patients memory deficits ch...   \n",
       "2     cglucose tissue co cglucose tissue electron ra...   \n",
       "3     pattern type thirteen patients type dat abilit...   \n",
       "4     blood rate bodies acids blood rate bodies acid...   \n",
       "...                                                 ...   \n",
       "3995  major genes influence outcome null alleles ass...   \n",
       "3996  cohort patients course patients years clinical...   \n",
       "3997  main classifications purpose paper trace descr...   \n",
       "3998  virus cell antigen pathway recombinant cell an...   \n",
       "3999  progression usefulness lymphocyte investigate ...   \n",
       "\n",
       "                                               BOAW_SFS  \\\n",
       "0     reduced amounts cortex brain tissue patients d...   \n",
       "1     criteria diagnosis patients presenting memory ...   \n",
       "2     synthesis cglucose tissue synthesis co cglucos...   \n",
       "3     pattern observations thirteen patients dat abi...   \n",
       "4     cerebral blood rate bodies amino acids cerebra...   \n",
       "...                                                 ...   \n",
       "3995  major genes influence outcome alleles explain ...   \n",
       "3996  cohort patients patients aged clinical examina...   \n",
       "3997  main classifications purpose paper trace descr...   \n",
       "3998  virus inhibit cell antigen pathway recombinant...   \n",
       "3999  predicting progression combined cd lymphocyte ...   \n",
       "\n",
       "                                                BON_SFS  \\\n",
       "0     amounts temporal cortex type brain tissue pati...   \n",
       "1     criteria type diagnosis patients memory charac...   \n",
       "2     synthesis cglucose tissue synthesis co cglucos...   \n",
       "3     pattern type observations patients type dat co...   \n",
       "4     cerebral blood rate bodies amino acids cerebra...   \n",
       "...                                                 ...   \n",
       "3995  major complex genes outcome null alleles assoc...   \n",
       "3996  cohort patients course patients clinical exami...   \n",
       "3997  main purpose paper trace describe principal cl...   \n",
       "3998  virus inhibit cell antigen pathway recombinant...   \n",
       "3999  progression usefulness cd lymphocyte purpose u...   \n",
       "\n",
       "                                               BONA_SFS  \\\n",
       "0     reduced amounts temporal cortex type brain tis...   \n",
       "1     criteria diagnosis patients memory deficits ch...   \n",
       "2     cglucose tissue co cglucose tissue electron ra...   \n",
       "3     pattern type thirteen patients type dat abilit...   \n",
       "4     cerebral blood rate bodies acids cerebral bloo...   \n",
       "...                                                 ...   \n",
       "3995  major genes influence outcome null alleles ass...   \n",
       "3996  cohort patients course patients years clinical...   \n",
       "3997  main classifications purpose paper trace descr...   \n",
       "3998  virus cell antigen pathway recombinant cell an...   \n",
       "3999  progression usefulness cd lymphocyte investiga...   \n",
       "\n",
       "                                              BOAW_TFRF  \\\n",
       "0     reduced amounts tissue patients diagnosis comp...   \n",
       "1     criteria diagnosis patients presenting memory ...   \n",
       "2     synthesis dioxide cglucose tissue synthesis co...   \n",
       "3     pattern observations thirteen patients dat abi...   \n",
       "4     blood rate bodies amino acids blood rate bodie...   \n",
       "...                                                 ...   \n",
       "3995  major genes influence outcome alleles explain ...   \n",
       "3996  cohort patients patients aged years clinical e...   \n",
       "3997  main classifications purpose paper trace descr...   \n",
       "3998  inhibit cell antigen pathway interference reco...   \n",
       "3999  predicting progression combined usefulness lym...   \n",
       "\n",
       "                                               BON_TFRF  \\\n",
       "0     amounts temporal type tissue patients diagnosi...   \n",
       "1     criteria type diagnosis patients memory defici...   \n",
       "2     synthesis dioxide production cglucose tissue s...   \n",
       "3     pattern reading type observations patients typ...   \n",
       "4     blood rate bodies amino acids blood rate bodie...   \n",
       "...                                                 ...   \n",
       "3995  major complex genes outcome null alleles diver...   \n",
       "3996  cohort patients course patients clinical exami...   \n",
       "3997  main purpose paper trace describe principal cl...   \n",
       "3998  immunodeficiency virus inhibit cell antigen pa...   \n",
       "3999  progression usefulness lymphocyte purpose usef...   \n",
       "\n",
       "                                              BONA_TFRF  \n",
       "0     reduced amounts temporal type tissue patients ...  \n",
       "1     criteria diagnosis patients memory deficits ch...  \n",
       "2     dioxide production cglucose tissue co producti...  \n",
       "3     pattern type thirteen patients type dat abilit...  \n",
       "4     blood rate bodies acids blood rate bodies acid...  \n",
       "...                                                 ...  \n",
       "3995  major complex genes influence outcome null all...  \n",
       "3996  cohort patients course patients years clinical...  \n",
       "3997  main classifications purpose paper trace descr...  \n",
       "3998  virus cell antigen pathway recombinant cell an...  \n",
       "3999  progression usefulness lymphocyte investigate ...  \n",
       "\n",
       "[4000 rows x 11 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read csv\n",
    "pubmedSelected = pd.read_csv('pubmed/selected.csv')\n",
    "scopusSelected = pd.read_csv('scopus/selected.csv')\n",
    "\n",
    "# get feature\n",
    "# pubmed\n",
    "pubmedClass = pubmedSelected.loc[:, 'class']\n",
    "pubmed_boaw_sr = pubmedSelected.loc[:, 'BOAW_SR']\n",
    "pubmed_bon_sr = pubmedSelected.loc[:, 'BON_SR']\n",
    "pubmed_bona_sr = pubmedSelected.loc[:, 'BONA_SR']\n",
    "pubmed_boaw_sfs = pubmedSelected.loc[:, 'BOAW_SFS']\n",
    "pubmed_bon_sfs = pubmedSelected.loc[:, 'BON_SFS']\n",
    "pubmed_bona_sfs = pubmedSelected.loc[:, 'BONA_SFS']\n",
    "pubmed_boaw_tfrf = pubmedSelected.loc[:, 'BOAW_TFRF']\n",
    "pubmed_bon_tfrf = pubmedSelected.loc[:, 'BON_TFRF']\n",
    "pubmed_bona_tfrf = pubmedSelected.loc[:, 'BONA_TFRF']\n",
    "\n",
    "# scopus\n",
    "scopusClass = scopusSelected.loc[:, 'class']\n",
    "scopus_boaw_sr = scopusSelected.loc[:, 'BOAW_SR']\n",
    "scopus_bon_sr = scopusSelected.loc[:, 'BON_SR']\n",
    "scopus_bona_sr = scopusSelected.loc[:, 'BONA_SR']\n",
    "scopus_boaw_sfs = scopusSelected.loc[:, 'BOAW_SFS']\n",
    "scopus_bon_sfs = scopusSelected.loc[:, 'BON_SFS']\n",
    "scopus_bona_sfs = scopusSelected.loc[:, 'BONA_SFS']\n",
    "scopus_boaw_tfrf = scopusSelected.loc[:, 'BOAW_TFRF']\n",
    "scopus_bon_tfrf = scopusSelected.loc[:, 'BON_TFRF']\n",
    "scopus_bona_tfrf = scopusSelected.loc[:, 'BONA_TFRF']\n",
    "pubmedSelected"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Term Weighting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# TF\n",
    "def tf(features):\n",
    "    # get tf weights & transform to sparse matrix\n",
    "    tfVec = CountVectorizer()\n",
    "    tf = tfVec.fit_transform(features.fillna(' '))\n",
    "    return tf\n",
    "\n",
    "# tfx = tf(bona)\n",
    "# print(tfx)\n",
    "\n",
    "# TF-IDF\n",
    "def tf_idf(features):\n",
    "    # get tf-idf weights & transform to sparse matrix\n",
    "    tfIdfVec = TfidfVectorizer()\n",
    "    tfIdf = tfIdfVec.fit_transform(features.fillna(' '))\n",
    "    return tfIdf\n",
    "\n",
    "# idfx = tf_idf(scopusBonSr)\n",
    "# print(idfx)\n",
    "\n",
    "# TF-IDF-ICF\n",
    "def tf_idf_icf(features, classes):\n",
    "    # count tf-idf\n",
    "    tfIdf = tf_idf(features)\n",
    "\n",
    "    # get classes weights\n",
    "    clsVec = LabelBinarizer()\n",
    "    cls = clsVec.fit_transform(classes)\n",
    "\n",
    "    # count total class\n",
    "    totalClass = cls.shape[1]\n",
    "\n",
    "    # count class contain feature (1 if contain, then convert to float, and sum per class)\n",
    "    classFeature = ((cls.T * tfIdf) > 0).astype(np.float64).sum(axis=0)\n",
    "\n",
    "    # count icf\n",
    "    icf = []\n",
    "    for featureId in range(tfIdf.shape[1]):\n",
    "        icf.append(1 + math.log(totalClass / classFeature[featureId]))\n",
    "    icf = np.array(icf)\n",
    "\n",
    "    # get tf-idf-icf weights & transform to sparse matrix\n",
    "    totalFeature = icf.shape[0]\n",
    "    tfIdfIcf = tfIdf * spdiags(icf, 0, totalFeature, totalFeature)\n",
    "    return tfIdfIcf\n",
    "\n",
    "# icfx = tf_idf_icf(pubmedBonSr, pubmedClass)\n",
    "# print(icfx)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# term weighting\n",
    "\n",
    "# pubmed\n",
    "# BOAW > SR > TF\n",
    "pubmed_boaw_sr_tf = tf(pubmed_boaw_sr)\n",
    "# BOAW > SR > TF-IDF\n",
    "pubmed_boaw_sr_tfIdf = tf_idf(pubmed_boaw_sr)\n",
    "# BOAW > SR > TF-IDF-ICF\n",
    "pubmed_boaw_sr_tfIdfIcf = tf_idf_icf(pubmed_boaw_sr, pubmedClass)\n",
    "# BOAW > TF-IDF\n",
    "pubmed_boaw_tfIdf = tf_idf(pubmed_boaw)\n",
    "# BOAW > TF-IDF-ICF\n",
    "pubmed_boaw_tfIdfIcf = tf_idf_icf(pubmed_boaw, pubmedClass)\n",
    "\n",
    "# BON > SR > TF\n",
    "pubmed_bon_sr_tf = tf(pubmed_bon_sr)\n",
    "# BON > SR > TF-IDF\n",
    "pubmed_bon_sr_tfIdf = tf_idf(pubmed_bon_sr)\n",
    "# BON > SR > TF-IDF-ICF\n",
    "pubmed_bon_sr_tfIdfIcf = tf_idf_icf(pubmed_bon_sr, pubmedClass)\n",
    "# BON > TF-IDF\n",
    "pubmed_bon_tfIdf = tf_idf(pubmed_bon)\n",
    "# BON > TF-IDF-ICF\n",
    "pubmed_bon_tfIdfIcf = tf_idf_icf(pubmed_bon, pubmedClass)\n",
    "\n",
    "# BONA > SR > TF\n",
    "pubmed_bona_sr_tf = tf(pubmed_bona_sr)\n",
    "# BONA > SR > TF-IDF\n",
    "pubmed_bona_sr_tfIdf = tf_idf(pubmed_bona_sr)\n",
    "# BONA > SR > TF-IDF-ICF\n",
    "pubmed_bona_sr_tfIdfIcf = tf_idf_icf(pubmed_bona_sr, pubmedClass)\n",
    "# BONA > TF-IDF\n",
    "pubmed_bona_tfIdf = tf_idf(pubmed_bona)\n",
    "# BONA > TF-IDF-ICF\n",
    "pubmed_bona_tfIdfIcf = tf_idf_icf(pubmed_bona, pubmedClass)\n",
    "\n",
    "\n",
    "# scopus\n",
    "# BOAW > SR > TF\n",
    "scopus_boaw_sr_tf = tf(scopus_boaw_sr)\n",
    "# BOAW > SR > TF-IDF\n",
    "scopus_boaw_sr_tfIdf = tf_idf(scopus_boaw_sr)\n",
    "# BOAW > SR > TF-IDF-ICF\n",
    "scopus_boaw_sr_tfIdfIcf = tf_idf_icf(scopus_boaw_sr, scopusClass)\n",
    "# BOAW > TF-IDF\n",
    "scopus_boaw_tfIdf = tf_idf(scopus_boaw)\n",
    "# BOAW > TF-IDF-ICF\n",
    "scopus_boaw_tfIdfIcf = tf_idf_icf(scopus_boaw, scopusClass)\n",
    "\n",
    "# BON > SR > TF\n",
    "scopus_bon_sr_tf = tf(scopus_bon_sr)\n",
    "# BON > SR > TF-IDF\n",
    "scopus_bon_sr_tfIdf = tf_idf(scopus_bon_sr)\n",
    "# BON > SR > TF-IDF-ICF\n",
    "scopus_bon_sr_tfIdfIcf = tf_idf_icf(scopus_bon_sr, scopusClass)\n",
    "# BON > TF-IDF\n",
    "scopus_bon_tfIdf = tf_idf(scopus_bon)\n",
    "# BON > TF-IDF-ICF\n",
    "scopus_bon_tfIdfIcf = tf_idf_icf(scopus_bon, scopusClass)\n",
    "\n",
    "# BONA > SR > TF\n",
    "scopus_bona_sr_tf = tf(scopus_bona_sr)\n",
    "# BONA > SR > TF-IDF\n",
    "scopus_bona_sr_tfIdf = tf_idf(scopus_bona_sr)\n",
    "# BONA > SR > TF-IDF-ICF\n",
    "scopus_bona_sr_tfIdfIcf = tf_idf_icf(scopus_bona_sr, scopusClass)\n",
    "# BONA > TF-IDF\n",
    "scopus_bona_tfIdf = tf_idf(scopus_bona)\n",
    "# BONA > TF-IDF-ICF\n",
    "scopus_bona_tfIdfIcf = tf_idf_icf(scopus_bona, scopusClass)\n",
    "# pubmed_bona_sr_tfIdfIcf"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Clustering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# K-Means++\n",
    "def kmeans_plus(data):\n",
    "    label = []\n",
    "    for k in tqdm(range(2, 11)):\n",
    "        kmp = KMeans(n_clusters=k, init = 'k-means++')\n",
    "        kmp.fit(data)\n",
    "        label.append(kmp.labels_)\n",
    "    return label\n",
    "\n",
    "# kmpx = kmeansPlus(pubmed_bona_sr_tfIdfIcf)\n",
    "# print(kmpx)\n",
    "\n",
    "# # Spherical K-Means\n",
    "# def spher_kmeans(data):\n",
    "#     label = []\n",
    "#     for k in tqdm(range(2, 11)):\n",
    "#         skm = SphericalKMeans(n_clusters=k)\n",
    "#         skm.fit(data)\n",
    "#         label = skm.labels_\n",
    "\n",
    "# skmx = spherKmeans(3, bona_sr_tfidf)\n",
    "# print(skmx)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  0%|          | 0/9 [00:00<?, ?it/s]C:\\Users\\ZephZ\\AppData\\Roaming\\Python\\Python311\\site-packages\\sklearn\\cluster\\_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning\n",
      "  warnings.warn(\n",
      "100%|██████████| 9/9 [00:07<00:00,  1.21it/s]\n",
      "100%|██████████| 9/9 [00:09<00:00,  1.05s/it]\n",
      "100%|██████████| 9/9 [00:08<00:00,  1.12it/s]\n",
      "100%|██████████| 9/9 [00:23<00:00,  2.63s/it]\n",
      "100%|██████████| 9/9 [00:17<00:00,  1.90s/it]\n"
     ]
    }
   ],
   "source": [
    "# get K-Means++ clusters\n",
    "# # pubmed\n",
    "# # BOAW > SR > TF > K-Means++\n",
    "# pubmed_boaw_sr_tf_kmeansPlus = kmeans_plus(pubmed_boaw_sr_tf)\n",
    "# # BOAW > SR > TF-IDF > K-Means++\n",
    "# pubmed_boaw_sr_tfIdf_kmeansPlus = kmeans_plus(pubmed_boaw_sr_tfIdf)\n",
    "# # BOAW > SR > TF-IDF-ICF > K-Means++\n",
    "# pubmed_boaw_sr_tfIdfIcf_kmeansPlus = kmeans_plus(pubmed_boaw_sr_tfIdfIcf)\n",
    "# # BOAW > TF-IDF > K-Means++\n",
    "# pubmed_boaw_tfIdf_kmeansPlus = kmeans_plus(pubmed_boaw_tfIdf)\n",
    "# # BOAW > TF-IDF-ICF > K-Means++\n",
    "# pubmed_boaw_tfIdfIcf_kmeansPlus = kmeans_plus(pubmed_boaw_tfIdfIcf)\n",
    "\n",
    "# BON > SR > TF > K-Means++\n",
    "pubmed_bon_sr_tf_kmeansPlus = kmeans_plus(pubmed_bon_sr_tf)\n",
    "# BON > SR > TF-IDF > K-Means++\n",
    "pubmed_bon_sr_tfIdf_kmeansPlus = kmeans_plus(pubmed_bon_sr_tfIdf)\n",
    "# BON > SR > TF-IDF-ICF > K-Means++\n",
    "pubmed_bon_sr_tfIdfIcf_kmeansPlus = kmeans_plus(pubmed_bon_sr_tfIdfIcf)\n",
    "# BON > TF-IDF > K-Means++\n",
    "pubmed_bon_tfIdf_kmeansPlus = kmeans_plus(pubmed_bon_tfIdf)\n",
    "# BON > TF-IDF-ICF > K-Means++\n",
    "pubmed_bon_tfIdfIcf_kmeansPlus = kmeans_plus(pubmed_bon_tfIdfIcf)\n",
    "\n",
    "# # BONA > SR > TF > K-Means++\n",
    "# pubmed_bona_sr_tf_kmeansPlus = kmeans_plus(pubmed_bona_sr_tf)\n",
    "# # BONA > SR > TF-IDF > K-Means++\n",
    "# pubmed_bona_sr_tfIdf_kmeansPlus = kmeans_plus(pubmed_bona_sr_tfIdf)\n",
    "# # BONA > SR > TF-IDF-ICF > K-Means++\n",
    "# pubmed_bona_sr_tfIdfIcf_kmeansPlus = kmeans_plus(pubmed_bona_sr_tfIdfIcf)\n",
    "# # BONA > TF-IDF > K-Means++\n",
    "# pubmed_bona_tfIdf_kmeansPlus = kmeans_plus(pubmed_bona_tfIdf)\n",
    "# # BONA > TF-IDF-ICF > K-Means++\n",
    "# pubmed_bona_tfIdfIcf_kmeansPlus = kmeans_plus(pubmed_bona_tfIdfIcf)\n",
    "\n",
    "\n",
    "# # scopus\n",
    "# # BOAW > SR > TF > K-Means++\n",
    "# scopus_boaw_sr_tf_kmeansPlus = kmeans_plus(scopus_boaw_sr_tf)\n",
    "# # BOAW > SR > TF-IDF > K-Means++\n",
    "# scopus_boaw_sr_tfIdf_kmeansPlus = kmeans_plus(scopus_boaw_sr_tfIdf)\n",
    "# # BOAW > SR > TF-IDF-ICF > K-Means++\n",
    "# scopus_boaw_sr_tfIdfIcf_kmeansPlus = kmeans_plus(scopus_boaw_sr_tfIdfIcf)\n",
    "# # BOAW > TF-IDF > K-Means++\n",
    "# scopus_boaw_tfIdf_kmeansPlus = kmeans_plus(scopus_boaw_tfIdf)\n",
    "# # BOAW > TF-IDF-ICF > K-Means++\n",
    "# scopus_boaw_tfIdfIcf_kmeansPlus = kmeans_plus(scopus_boaw_tfIdfIcf)\n",
    "\n",
    "# # BON > SR > TF > K-Means++\n",
    "# scopus_bon_sr_tf_kmeansPlus = kmeans_plus(scopus_bon_sr_tf)\n",
    "# # BON > SR > TF-IDF > K-Means++\n",
    "# scopus_bon_sr_tfIdf_kmeansPlus = kmeans_plus(scopus_bon_sr_tfIdf)\n",
    "# # BON > SR > TF-IDF-ICF > K-Means++\n",
    "# scopus_bon_sr_tfIdfIcf_kmeansPlus = kmeans_plus(scopus_bon_sr_tfIdfIcf)\n",
    "# # BON > TF-IDF > K-Means++\n",
    "# scopus_bon_tfIdf_kmeansPlus = kmeans_plus(scopus_bon_tfIdf)\n",
    "# # BON > TF-IDF-ICF > K-Means++\n",
    "# scopus_bon_tfIdfIcf_kmeansPlus = kmeans_plus(scopus_bon_tfIdfIcf)\n",
    "\n",
    "# # BONA > SR > TF > K-Means++\n",
    "# scopus_bona_sr_tf_kmeansPlus = kmeans_plus(scopus_bona_sr_tf)\n",
    "# # BONA > SR > TF-IDF > K-Means++\n",
    "# scopus_bona_sr_tfIdf_kmeansPlus = kmeans_plus(scopus_bona_sr_tfIdf)\n",
    "# # BONA > SR > TF-IDF-ICF > K-Means++\n",
    "# scopus_bona_sr_tfIdfIcf_kmeansPlus = kmeans_plus(scopus_bona_sr_tfIdfIcf)\n",
    "# # BONA > TF-IDF > K-Means++\n",
    "# scopus_bona_tfIdf_kmeansPlus = kmeans_plus(scopus_bona_tfIdf)\n",
    "# # BONA > TF-IDF-ICF > K-Means++\n",
    "# scopus_bona_tfIdfIcf_kmeansPlus = kmeans_plus(scopus_bona_tfIdfIcf)\n",
    "# # pubmed_bona_sr_tfIdfIcf_kmeansPlus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get Spherical K-Means clusters\n",
    "# pubmed\n",
    "\n",
    "# scopus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# save json\n",
    "# pubmed\n",
    "# pubmedJson = {\n",
    "#     'boaw_sr_tf_kmeansPlus': [],\n",
    "    # 'boaw_sr_tfIdf_kmeansPlus': pubmed_boaw_sr_tfIdf_kmeansPlus,\n",
    "    # 'boaw_sr_tfIdfIcf_kmeansPlus': pubmed_boaw_sr_tfIdfIcf_kmeansPlus,\n",
    "    # 'boaw_tfIdf_kmeansPlus': pubmed_boaw_tfIdf_kmeansPlus,\n",
    "    # 'boaw_tfIdfIcf_kmeansPlus': pubmed_boaw_tfIdfIcf_kmeansPlus,\n",
    "    # 'bon_sr_tf_kmeansPlus': pubmed_bon_sr_tf_kmeansPlus,\n",
    "    # 'bon_sr_tfIdf_kmeansPlus': pubmed_bon_sr_tfIdf_kmeansPlus,\n",
    "    # 'bon_sr_tfIdfIcf_kmeansPlus': pubmed_bon_sr_tfIdfIcf_kmeansPlus,\n",
    "    # 'bon_tfIdf_kmeansPlus': pubmed_bon_tfIdf_kmeansPlus,\n",
    "    # 'bon_tfIdfIcf_kmeansPlus': pubmed_bon_tfIdfIcf_kmeansPlus,\n",
    "    # 'bona_sr_tf_kmeansPlus': pubmed_bona_sr_tf_kmeansPlus,\n",
    "    # 'bona_sr_tfIdf_kmeansPlus': pubmed_bona_sr_tfIdf_kmeansPlus,\n",
    "    # 'bona_sr_tfIdfIcf_kmeansPlus': pubmed_bona_sr_tfIdfIcf_kmeansPlus,\n",
    "    # 'bona_tfIdf_kmeansPlus': pubmed_bona_tfIdf_kmeansPlus,\n",
    "    # 'bona_tfIdfIcf_kmeansPlus': pubmed_bona_tfIdfIcf_kmeansPlus\n",
    "    # spher_kmeans\n",
    "# }\n",
    "# pubmedJson['boaw_sr_tf_kmeansPlus'].append(pubmed_boaw_sr_tf_kmeansPlus)\n",
    "# with open('pubmed/clustered.json', 'w', encoding='utf-8') as f:\n",
    "#     json.dump(pubmedJson, f, indent=4)\n",
    "# f.close() \n",
    "\n",
    "# scopus\n",
    "# scopus_boaw_sr_tf_kmeansPlus,\n",
    "# scopus_boaw_sr_tfIdf_kmeansPlus,\n",
    "# scopus_boaw_sr_tfIdfIcf_kmeansPlus,\n",
    "# scopus_boaw_tfIdf_kmeansPlus,\n",
    "# scopus_boaw_tfIdfIcf_kmeansPlus,\n",
    "# scopus_bon_sr_tf_kmeansPlus,\n",
    "# scopus_bon_sr_tfIdf_kmeansPlus,\n",
    "# scopus_bon_sr_tfIdfIcf_kmeansPlus\n",
    "# scopus_bon_tfIdf_kmeansPlus,\n",
    "# scopus_bon_tfIdfIcf_kmeansPlus,\n",
    "# scopus_bona_sr_tf_kmeansPlus,\n",
    "# scopus_bona_sr_tfIdf_kmeansPlus,\n",
    "# scopus_bona_sr_tfIdfIcf_kmeansPlus,\n",
    "# scopus_bona_tfIdf_kmeansPlus,\n",
    "# scopus_bona_tfIdfIcf_kmeansPlus,"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# read json\n",
    "# pubmed\n",
    "\n",
    "# scopus\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Performance Analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# encode class as label number\n",
    "def encodeClass(classes):\n",
    "    # encode class\n",
    "    clsVec = LabelEncoder()\n",
    "    cls = clsVec.fit_transform(classes)\n",
    "    # add 1 to all list (to compare with label)\n",
    "    return cls + 1\n",
    "pubmedLabel = encodeClass(pubmedClass)\n",
    "scopusLabel = encodeClass(scopusClass)\n",
    "\n",
    "# Silhouette Score\n",
    "def silhouette(documentWeight, clusterLabel):\n",
    "    score = []\n",
    "    for k in tqdm(range(len(clusterLabel))):\n",
    "        silhouetteScore = silhouette_score(documentWeight, clusterLabel[k])\n",
    "        score.append(silhouetteScore)\n",
    "    return score\n",
    "\n",
    "# sx = silhouette(pubmed_bona_sr_tfIdfIcf, kmpx)\n",
    "# print(sx)\n",
    "\n",
    "# Purity\n",
    "def purity(documentLabel, clusterLabel):\n",
    "    score = []\n",
    "    for k in tqdm(range(len(clusterLabel))):\n",
    "        contingencyMatrix = contingency_matrix(documentLabel, clusterLabel[k])\n",
    "        purityScore = np.sum(np.amax(contingencyMatrix, axis=0)) / np.sum(contingencyMatrix)\n",
    "        score.append(purityScore)\n",
    "    return score\n",
    "\n",
    "# px = purity(pubmedLabel, kmpx)\n",
    "# print(px)\n",
    "\n",
    "# AMI\n",
    "def ami(documentLabel, clusterLabel):\n",
    "    score = []\n",
    "    for k in tqdm(range(len(clusterLabel))):\n",
    "        amiScore = adjusted_mutual_info_score(documentLabel, clusterLabel[k])\n",
    "        score.append(amiScore)\n",
    "    return score\n",
    "\n",
    "# ax = ami(pubmedLabel, kmpx)\n",
    "# print(ax)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 9/9 [00:04<00:00,  1.85it/s]\n"
     ]
    }
   ],
   "source": [
    "sr_tfIdfIcf = silhouette(pubmed_bon_sr_tfIdfIcf, pubmed_bon_sr_tfIdfIcf_kmeansPlus)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 9/9 [00:05<00:00,  1.59it/s]\n"
     ]
    }
   ],
   "source": [
    "tfIdfIcf = silhouette(pubmed_bon_tfIdfIcf, pubmed_bon_tfIdfIcf_kmeansPlus)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.0025779042220132027"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "max(sr_tfIdfIcf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.01554582499538646"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "max(tfIdfIcf)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.2"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "ab696253bf1cbd9262a120019f89a6af5e719602af9132e0a92b18d085125844"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
